### PR TITLE
TaskWrapper and simplication of Tasks and JobResults

### PR DESCRIPTION
**Pull request by dbarrosop**
This PR opens the discussion for simplifying a few things:

1. First, it changes the Task interface to:

``` golang
 type Task interface {
	Run(context.Context, *Host) (interface{}, error)
}
```

So now `Task.Run` is just a regular function that only cares about doing A on host. No need to remember calling `wg.Done()` and no need to return stuff via the channel. Instead, just do your thing and return whatever you want and an error.

2. The runners now call the task via a wrapper function `TaskWrapper` that serves as decorator, this is where we do now everything that is related to the parallelism; call `wg.Done()`, wrap the result in a `JobResult` object and pass it via the channel.

3. Get rid of `JobParameters" which didn't seem to server a very useful purpose and made things a bit more confusing than necessary.

This achieves a few things:

1. Tasks are now simpler as they only need to care about themselves
2. Tasks are also easier to test as there are less things to pass to them
3. We control in the wrapper what we want to do, for instance, we call `jp.Host().setErr(err)` to store the error in the Host. We may want to log as well or do some other stuff in the future, point is, we can do it here instead of documenting "expectations".

I have also removed a couple of attributes from the `JobResult`; `changed` and `SubResults`. Not sure at this point if `changed` is needed so I just removed it thinking we can add it later if needed. `SubResults` is removed in favor of just letting the tasks handle that themselves if needed, you can find an example [here](https://github.com/nornir-automation/gornir/compare/test_runners...taskWrapper?expand=1#diff-8ab6487618252b8557c0a5194fa64f49R23)

Finally, `output.RenderResults` instead of inspecting objects and trying to do magic calls `fmt.Sprintf`, which means tasks results can implement the function `String` and provide their own formatting. For [example](https://github.com/nornir-automation/gornir/compare/test_runners...taskWrapper?expand=1#diff-8ab6487618252b8557c0a5194fa64f49R26).

It's still missing cleaning, updating documentation, and adding tests but wanted to open it to show the progress and kick the discussion.

### CODE REVIEW

**Comment by nleiva** [comment]
Sounds good, I'm in favor of making things as simple as possible. I'll let you develop a bit more on the code before submitting reviews.

One thing that noticed in another PR and we might also want to consider here is that Go will run goroutines concurrently, which is not the same as in parallel. You will effectively run them in parallel only if you have the resources at your disposal, which is not always the case.

**Comment by dbarrosop** [comment]
Ok, this one is ready. I think the result is quite good as we maintain the same functionality removing a bunch of stuff and simplifying dramatically the whole development of tasks.

I know it looks like a lot because it touches 23 files but most of the changes are superfluous. Let me know if you have questions or comments :)

File: pkg/gornir/runner.go [hunk 1]
```diff
@@ -5,69 +5,39 @@ import (
...
+	Run(context.Context, Logger, *Host) (TaskInstanceResult, error)
 }
 
 // Runner is the interface of a struct that can implement a strategy
 // to run tasks over hosts
 type Runner interface {
-	Run(context.Context, Task, map[string]*Host, *JobParameters, chan *JobResult) error // Run executes the task over the hosts
-	Close() error                                                                       // Close closes and cleans all objects associated with the runner
-	Wait() error                                                                        // Wait blocks until all the hosts are done executing the task
+	Run(context.Context, Logger, Task, map[string]*Host, chan *JobResult) error // Run executes the task over the hosts
```
Comment (nleiva) [inline pkg/gornir/runner.go]: Isn't `Logger` part of the Gornir?. You are also passing an `Inventory` (`map[string]*Host`), which is also in Gornir. I'm wondering if there is a cleaner way to pass all these values around. Probably something we can discuss in another PR or Issue.

File: pkg/gornir/runner.go [hunk 2]
```diff
@@ -5,69 +5,39 @@ import (
...
-}
+// TaskInstanceResult is the resut of running a task on a given host
+type TaskInstanceResult interface{}
 
 // Task is the interface that task plugins need to implement.
 // the task is responsible to indicate its completion
 // by calling sync.WaitGroup.Done()
 type Task interface {
-	Run(context.Context, *sync.WaitGroup, *JobParameters, chan *JobResult)
+	Run(context.Context, Logger, *Host) (TaskInstanceResult, error)
```
Comment (nleiva) [inline pkg/gornir/runner.go]: I think we could change the name of `Run` either here or at the Runner to make the workflow easier to follow. So when I see a Run, I immediately know if we are talking about a Task OR a Runner (I know, they have different signatures, but still)

So, move from "The Runner **Run** the Task that is **Run**" to "The Runner **Run** the Task that it **Execute**" or "The Runner **Trigger** the Task that is **Run**" or something else.

File: Dockerfile [hunk 3]
```diff
@@ -4,4 +4,4 @@ ENV GO111MODULE=on
 ENV GOPROXY=https://proxy.golang.org
 WORKDIR /go/src/github.com/nornir-automation/gornir
 ADD . .
-RUN go mod download
+RUN go mod download > /dev/null
```
[the remark this answers is not in the document]
Comment (dbarrosop) [reply Dockerfile]: Mmm, I actually don't remember why I added the redirection. Removing and if tests break I will try to remember what that was about

[the remark this answers is not in the document]
Comment (dbarrosop) [reply Dockerfile]: Ok, just removed this before merging.

**Comment by dbarrosop** [comment]
Thanks for the review!

**Pull request merged**